app/controllers/benefits/CompanyBenefitController.scala

```diff
@@ -113,20 +118,40 @@ class CompanyBenefitController @Inject()(
         }
       },
       success => {
-        success match {
-          case Some(NoIDontGetThisBenefit) =>
-            journeyCacheService.cache(DecisionChoice, NoIDontGetThisBenefit) map { _ =>
-              Redirect(controllers.benefits.routes.RemoveCompanyBenefitController.stopDate())
-            }
-          case Some(YesIGetThisBenefit) =>
-            journeyCacheService.cache(DecisionChoice, YesIGetThisBenefit) map { _ =>
-              Redirect(
-                controllers.routes.ExternalServiceRedirectController
-                  .auditInvalidateCacheAndRedirectService(TaiConstants.CompanyBenefitsIform)
-                  .url)
+        val benefitType = journeyCacheService.mandatoryJourneyValueAs(
+          EndCompanyBenefit_BenefitTypeKey,
+          x => x
+        )
+
+        benefitType.flatMap {
+          case Right(name) =>
+            success match {
+              case Some(NoIDontGetThisBenefit) =>
+                journeyCacheService
+                  .cache(getBenefitDecisionKey(Some(name), DecisionChoice), NoIDontGetThisBenefit)
+                  .map { _ =>
+                    Redirect(controllers.benefits.routes.RemoveCompanyBenefitController.stopDate())
+                  }
+              case Some(YesIGetThisBenefit) =>
+                journeyCacheService.cache(getBenefitDecisionKey(Some(name), DecisionChoice), YesIGetThisBenefit).map {
+                  _ =>
+                    Redirect(controllers.routes.ExternalServiceRedirectController
+                      .auditInvalidateCacheAndRedirectService(TaiConstants.CompanyBenefitsIform))
+                }
+              case e => {
+                logger.error(s"Data in incorrect format: $e")
+                Future.successful(Redirect(controllers.routes.TaxAccountSummaryController.onPageLoad()))
+              }
             }
+
+          case Left(e) => {
+            Future.successful(Redirect(controllers.routes.TaxAccountSummaryController.onPageLoad()))
+          }
         }
       }
     )
   }
+
+  def getBenefitDecisionKey(benefitType: Option[String], decision: String): String =
+    s"$benefitType $decision"
```

Comment:
Do you want to interpolate over an Option here? It'll return something like "None Decision" or "Some(benefitType) decision"

Reply:
No, Ideally I do want to have just "Telephone NoIDontGetThisBenefit", so would you recommend I fold benefit type so I get "Telephone NoIDontGetThisBenefit" if there is a value or "NoIDontGetThisBenefit" if there is no benefitType value?

Reply:
Mapping over benefitType first and ensure you always pass in a string into `getBenefitDecisionKey` is probably the way to go as you'd never call this function on a None

Reply:
getBenefitDecisionKey now has different behaviour in the decision function. It's obtaining the cache values on `$suffix` which is never cached, so you will always get a None back when calling `journeycacheService.getCurrentValue("suffix is never cached")`

It's misleading and can make it difficult to read and debug at a later date. I suggest making the `getBenefitDecisionKey` return an Option[String], this will add a bit of code complexity as you have to match on this option but it means the behaviour hasn't changed pre-refactor and you know you're trying to get a None from `journeycacheService.getCurrentValue` when you initially land on this page. 

To make it easier to read and to prevent someone accidentally changing the cached/extracted key, I would pull the caching logic to its own file.